numpy/_core/src/umath/loops_trigonometric.dispatch.cpp

```diff
@@ -0,0 +1,298 @@
+ * range [-71476.0625f, 71476.0625f] for cosine and [-117435.992f, 117435.992f]
+ * for sine.
+ * (2) For elements within range, perform range reduction using
+ * Cody-Waite's method: x* = x - y*PI/2, where y = rint(x*2/PI). x* \in [-PI/4,
+ * PI/4].
+ * (3) Map cos(x) to (+/-)sine or (+/-)cosine of x* based on the
+ * quadrant k = int(y).
+ * (4) For elements outside that range, Cody-Waite
+ * reduction performs poorly leading to catastrophic cancellation. We compute
+ * cosine by calling glibc in a scalar fashion.
+ * (5) Vectorized implementation
+ * has a max ULP of 1.49 and performs at least 5-7x(x86) - 2.5-3x(Power) -
+ * 1-2x(Arm) faster than scalar implementations when magnitude of all elements
+ * in the array < 71476.0625f (117435.992f for sine).  Worst case performance
+ * is when all the elements are large leading to about 1-2% reduction in
+ * performance.
+ * TODO: use vectorized version of Payne-Hanek style reduction for large
+ * elements or when there's no native FUSED support instead of fallback to libc
+ */
+
+#if NPY_SIMD_FMA3  // native support
+typedef enum
+{
+    SIMD_COMPUTE_SIN,
+    SIMD_COMPUTE_COS
+} SIMD_TRIG_OP;
+
+const hn::ScalableTag<float> f32;
+const hn::ScalableTag<int32_t> s32;
+using vec_f32 = hn::Vec<decltype(f32)>;
+using vec_s32 = hn::Vec<decltype(s32)>;
+using opmask_t = hn::Mask<decltype(f32)>;
+
+NPY_FINLINE vec_f32
+simd_range_reduction_f32(vec_f32 x, vec_f32 y, vec_f32 c1, vec_f32 c2, vec_f32 c3)
+{
+    vec_f32 reduced_x = hn::MulAdd(y, c1, x);
+    reduced_x = hn::MulAdd(y, c2, reduced_x);
+    reduced_x = hn::MulAdd(y, c3, reduced_x);
+    return reduced_x;
+}
+
+NPY_FINLINE vec_f32
+simd_cosine_poly_f32(vec_f32 x2)
+{
+    const vec_f32 invf8 = hn::Set(f32, 0x1.98e616p-16f);
+    const vec_f32 invf6 = hn::Set(f32, -0x1.6c06dcp-10f);
+    const vec_f32 invf4 = hn::Set(f32, 0x1.55553cp-05f);
+    const vec_f32 invf2 = hn::Set(f32, -0x1.000000p-01f);
+    const vec_f32 invf0 = hn::Set(f32, 0x1.000000p+00f);
+
+    vec_f32 r = hn::MulAdd(invf8, x2, invf6);
+    r = hn::MulAdd(r, x2, invf4);
+    r = hn::MulAdd(r, x2, invf2);
+    r = hn::MulAdd(r, x2, invf0);
+    return r;
+}
+/*
+ * Approximate sine algorithm for x \in [-PI/4, PI/4]
+ * Maximum ULP across all 32-bit floats = 0.647
+ * Polynomial approximation based on unpublished work by T. Myklebust
+ */
+NPY_FINLINE vec_f32
+simd_sine_poly_f32(vec_f32 x, vec_f32 x2)
+{
+    const vec_f32 invf9 = hn::Set(f32, 0x1.7d3bbcp-19f);
+    const vec_f32 invf7 = hn::Set(f32, -0x1.a06bbap-13f);
+    const vec_f32 invf5 = hn::Set(f32, 0x1.11119ap-07f);
+    const vec_f32 invf3 = hn::Set(f32, -0x1.555556p-03f);
+
+    vec_f32 r = hn::MulAdd(invf9, x2, invf7);
+    r = hn::MulAdd(r, x2, invf5);
+    r = hn::MulAdd(r, x2, invf3);
+    r = hn::MulAdd(r, x2, hn::Zero(f32));
+    r = hn::MulAdd(r, x, x);
+    return r;
+}
+
+NPY_FINLINE vec_f32
+GatherIndexN(const float* src, npy_intp ssrc, npy_intp len)
```

Comment:
I'm not massively familiar with x86, but it looks like `npyv_loadn_tillz_s32` uses a gather instruction here rather than a loop, should try it to see if it provides similar performance?

Reply:
The 1.5x slowdown actually happens in the non strided case which don't use Gather or Scatter. But even for the strided case, using gather is slower than scalar method (since the DOWNFALL CVE).

Reply:
Ahh, apologies, I wasn't sure if the code generated for Gather was pretty slow on x86 as well, it doesn't look as good as the ASIMD `npyv_` function 😅 Too many architectures 😸

Reply:
Why not use hn:: GatherIndexN? 
VI indics can be get by
hn::Iota(i32, 0) * ssrc